src/Dialect/ONNX/MLIRDialectBuilder.cpp

```diff
@@ -676,34 +697,50 @@ Value VectorBuilder::mergeLow(Value lhs, Value rhs, int64_t step) {
   return shuffle(lhs, rhs, mask);
 }
 
-// Composite functions
-Value VectorBuilder::multiReduction(SmallVectorImpl<Value> &vecArray) {
-  uint64_t N = vecArray.size();
+// Do a parallel-simd reduction of N vectors of SIMD length VL.
+// Restrictions:
+// *  VL is the vector length of the machine SIMD vectors.
+// *  N is a multiple of VL as we can perform consecutive VL x VL
+//    reductions.
+void VectorBuilder::multiReduction(SmallVectorImpl<Value> &inputVecArray,
+    SmallVectorImpl<Value> &outputVecArray) {
+  uint64_t N = inputVecArray.size();
   assert(N > 0 && "expected at least one value to reduce");
-  uint64_t VL = vector1DLength(vecArray[0]);
-  LLVM_DEBUG(
-      llvm::dbgs() << "reduction with N " << N << ", VL " << VL << "\n";);
-  assert(VL == N && "expected the same number of vectors in array as VL");
-  assert(VL == 4 && "only natural sizes supported at this time");
+  uint64_t VL = getLengthOf1DVector(inputVecArray[0]);
+  uint64_t machineVL = getMachineVectorLength(inputVecArray[0]);
+  assert(VL == machineVL && "only natural sizes supported at this time");
+  assert(N % machineVL == 0 &&
```

Comment:
[nit]:  Comments states that N must be a multiple of VL so would write `N % VL == 0` here (in case the restriction on machineVL is removed later on).

Reply:
It is the reverse, I may be able to support in the future VL being multiple of machineVL, this is why it is already future proof as is

Reply:
I see thanks !